### PR TITLE
[Feature] Support parallel scan on tablet for merge/agg table

### LINKED ISSUE #6110: Support parallel scan on the tablet
## Feature request

Support parallel scan on the tablet ONLY for the query.
- [ ] Physically split tablet by row id range.
  This could be used by the non-merge/aggregate table model, including `PRIMARY_KEYS`, `DUP_KEYS`, and `UNIQUE_KEYS` which skip aggreration.
- [ ] Logically split tablet by the key range.
    This could be used by the merge/aggregate table model.
- [ ] Use the memory occupation of scan rows to limit the DOP of the scan operator.

### PR DESCRIPTION
**Pull request by ZiheLiu**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR relates：
Closes #6110.

## Problem Summary：

As for parallel scan on the single tablet, it splits a key range into multiple small short key ranges. Add a new `LogicalSplitMorselQueue` to implement this.

## Modification
- `MorselQueue` creates multiple morsels from a tablet.
- As for a tablet, it splits each key range into small short key ranges.
- It uses short key blocks on the non-overlapping segments of the largest segment for the current tablet to split each key range.

### CODE REVIEW

File: be/src/exec/pipeline/scan/morsel.cpp [hunk 1]
```diff
@@ -229,5 +236,304 @@ Status PhysicalSplitMorselQueue::_init_segment() {
...
+        _range_end_key.emplace_back(key_range->end_scan_range);
+    }
+}
+
+StatusOr<MorselPtr> LogicalSplitMorselQueue::try_get() {
+    DCHECK(!_tablets.empty());
+    DCHECK(!_tablet_rowsets.empty());
+    DCHECK_EQ(_tablets.size(), _tablet_rowsets.size());
+
+    std::lock_guard<std::mutex> lock(_mutex);
```
[the remark this answers is not in the document]
Comment (ZiheLiu) [reply be/src/exec/pipeline/scan/morsel.cpp:262]: > Will this lock cause too much contention?

No, according to the test result, reading a morsel costs much more time then this picking up morsel.

[the remark this answers is not in the document]
Comment (ZiheLiu) [reply be/src/exec/pipeline/scan/morsel.cpp:262]: > Why other methods don't need this lock?

The other methods only return constant values except `empty()`.
`empty()` uses atomic `_tablet_idx` to indicate whether there are tablets not finished to be splitted.

File: be/src/storage/rowset/segment_group.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+Slice ShortKeyIndexGroupIterator::operator*() const {
+    DCHECK(valid());
+    return _decoder_group->key(_ordinal);
+}
+
+/// ShortKeyIndexDecoderGroup
+ShortKeyIndexDecoderGroup::ShortKeyIndexDecoderGroup(std::vector<const ShortKeyIndexDecoder*> sk_index_decoders)
+        : _sk_index_decoders(std::move(sk_index_decoders)) {
```
[the remark this answers is not in the document]
Comment (ZiheLiu) [reply be/src/storage/rowset/segment_group.cpp:22]: Great.

File: be/src/storage/rowset/segment_group.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+/// SegmentGroup
+SegmentGroup::SegmentGroup(std::vector<SegmentSharedPtr> segments) : _segments(std::move(segments)) {
+    std::vector<const ShortKeyIndexDecoder*> sk_index_decoders;
+    sk_index_decoders.reserve(_segments.size());
+    for (const auto& segment : _segments) {
+        DCHECK(segment->has_loaded_index());
+        sk_index_decoders.emplace_back(segment->decoder());
+    }
+    _decoder_group = std::make_unique<ShortKeyIndexDecoderGroup>(std::move(sk_index_decoders));
```
[the remark this answers is not in the document]
Comment (ZiheLiu) [reply be/src/storage/rowset/segment_group.cpp]: I changed the parameters of `ShortKeyIndexDecoderGroup`, and now `_decoder_group` could be the plain object.

File: be/src/exec/pipeline/scan/morsel.cpp [hunk 4]
```diff
@@ -229,5 +236,304 @@ Status PhysicalSplitMorselQueue::_init_segment() {
...
+        return std::make_unique<vectorized::ShortKeyOption>(_short_key_schema, short_key, false);
+    }
+}
+
+bool LogicalSplitMorselQueue::_cur_tablet_finished() const {
+    return _range_idx >= _block_ranges_per_seek_range.size();
+}
+
+Rowset* LogicalSplitMorselQueue::_find_largest_rowset(const std::vector<RowsetSharedPtr>& rowsets) {
+    Rowset* largest_rowset = nullptr;
```
Comment (kangkaisen) [inline be/src/exec/pipeline/scan/morsel.cpp:402]: Could assign the first rowset to largest_rowset, then we needn't largest_rowset == nullptr check?

File: be/src/storage/rowset/segment_group.h [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+private:
+    // The short keys among decoders are increasing.
+    std::vector<const ShortKeyIndexDecoder*> _sk_index_decoders;
+    // _decoder_start_ordinals[i] represents the start ordinal of _sk_index_decoder[i].
+    // _decoder_start_ordinals[_sk_index_decoders.size()] represents the total number of rows among decoders.
+    std::vector<ssize_t> _decoder_start_ordinals;
+};
+
+class SegmentGroup {
```
Comment (kangkaisen) [inline be/src/storage/rowset/segment_group.h:95]: need to comment what's the SegmentGroup, and why we need it.

File: be/src/storage/rowset/segment_group.h [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    Slice operator*() const;
+
+    ssize_t ordinal() const { return _ordinal; }
+
+private:
+    const ShortKeyIndexDecoderGroup* _decoder_group;
+    ssize_t _ordinal;
+};
+
+class ShortKeyIndexDecoderGroup {
```
Comment (kangkaisen) [inline be/src/storage/rowset/segment_group.h:67]: DecoderGroup sounds a little strange

File: be/src/storage/rowset/segment_group.cpp [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    ssize_t block_id_in_decoder;
+    _find_position(ordinal, &decoder_id, &block_id_in_decoder);
+
+    return _sk_index_decoders[size_t(decoder_id)]->key(block_id_in_decoder);
+}
+
+template <bool lower_bound>
+ShortKeyIndexGroupIterator ShortKeyIndexDecoderGroup::_seek(const Slice& key) const {
+    auto comparator = [](const Slice& lhs, const Slice& rhs) { return lhs.compare(rhs) < 0; };
+    if (lower_bound) {
```
Comment (kangkaisen) [inline be/src/storage/rowset/segment_group.cpp:70]: constexpr

File: be/src/storage/chunk_helper.h [hunk 8]
```diff
@@ -48,6 +48,9 @@ class ChunkHelper {
     static vectorized::Schema convert_schema_to_format_v2(const starrocks::TabletSchema& schema,
                                                           const std::vector<ColumnId>& cids);
 
+    // Get schema with format v2 type containing short key columns from TabletSchema.
+    static vectorized::Schema get_short_key_schema_with_format_v2(const starrocks::TabletSchema& schema);
```
Comment (kangkaisen) [inline be/src/storage/chunk_helper.h:52]: Currently, we are all format v2.

File: be/src/exec/pipeline/scan/morsel.h [hunk 9]
```diff
@@ -199,5 +220,83 @@ class PhysicalSplitMorselQueue final : public MorselQueue {
...
+    Morsels _morsels;
+    // The number of the morsels before split them to pieces.
+    const size_t _num_original_morsels;
+    const int64_t _degree_of_parallelism;
+    // The minimum number of rows picked up from a segment at one time.
+    const int64_t _splitted_scan_rows;
+
+    /// Key ranges passed to the storage layer.
+    // Possible values are "gt", "ge", "eq".
+    std::string _range_start_op = "";
```
Comment (kangkaisen) [inline be/src/exec/pipeline/scan/morsel.h:274]: could we use enum, not string?

File: be/src/exec/pipeline/scan/morsel.cpp (hunk 4 above)
Comment (ZiheLiu) [reply be/src/exec/pipeline/scan/morsel.cpp:402]: Great.

File: be/src/storage/rowset/segment_group.h (hunk 6 above)
Comment (ZiheLiu) [reply be/src/storage/rowset/segment_group.h:67]: This class contains `ShortKeyIndexDecoder` for each segment, and perform the binary search logic. Any better idea?

File: be/src/storage/chunk_helper.h (hunk 8 above)
Comment (ZiheLiu) [reply be/src/storage/chunk_helper.h:52]: It is to be consistent style with other methods, such as `convert_schema_to_format_v2`

File: be/src/exec/pipeline/scan/morsel.h (hunk 9 above)
Comment (ZiheLiu) [reply be/src/exec/pipeline/scan/morsel.h]: Done.

**Comment by ZiheLiu** [comment]
run starrocks_clang-format

File: be/src/exec/pipeline/scan/morsel.cpp [hunk 10]
```diff
@@ -229,5 +238,315 @@ Status PhysicalSplitMorselQueue::_init_segment() {
...
+        }
+        cur_num_taken_blocks = std::min(cur_num_taken_blocks, num_rest_blocks);
+
+        // As for the last key range, if there is not enough blocks after this time taking,
+        // take from the last key range at most half of the rest blocks.
+        if (_range_idx + 1 >= _block_ranges_per_seek_range.size() && num_rest_blocks > cur_num_taken_blocks &&
+            num_rest_blocks - cur_num_taken_blocks < _sample_splitted_scan_blocks) {
+            cur_num_taken_blocks = std::min(cur_num_taken_blocks, num_rest_blocks / 2);
+            need_more_blocks = false;
+        }C
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/scan/morsel.cpp:313]: What's this character?

Comment (ZiheLiu) [reply be/src/exec/pipeline/scan/morsel.cpp]: Oh, a typo, fixed.

File: be/src/exec/pipeline/scan/morsel.cpp [hunk 11]
```diff
@@ -229,5 +238,315 @@ Status PhysicalSplitMorselQueue::_init_segment() {
...
+
+    if (_tablet_idx == 0) {
+        // All the tablets have the same schema, so parse seek range with the first table schema.
+        RETURN_IF_ERROR(vectorized::TabletReader::parse_seek_range(_tablets[_tablet_idx], _range_start_op,
+                                                                   _range_end_op, _range_start_key, _range_end_key,
+                                                                   &_tablet_seek_ranges, &_mempool));
+    }
+
+    _largest_rowset = _find_largest_rowset(_tablet_rowsets[_tablet_idx]);
+    if (_largest_rowset == nullptr || _largest_rowset->num_rows() == 0) {
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/scan/morsel.cpp:479]: Is it safe to ignore the initialization of the following fields when no rowset exist?

File: be/src/exec/pipeline/scan/morsel.cpp [hunk 12]
```diff
@@ -229,5 +238,315 @@ Status PhysicalSplitMorselQueue::_init_segment() {
...
+        }
+        RETURN_IF_ERROR(_init_tablet());
+    }
+
+    size_t num_taken_blocks = 0;
+    std::vector<vectorized::ShortKeyRangeOptionPtr> short_key_ranges;
+    vectorized::ShortKeyOptionPtr _cur_range_lower = nullptr;
+    vectorized::ShortKeyOptionPtr _cur_range_upper = nullptr;
+    bool need_more_blocks = true;
+    while (!_cur_tablet_finished() &&      // One morsel only read data from one tablet.
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/scan/morsel.cpp:288]: Had better to give some examples to illustrate the detail

**Review by murphyatwork — DISMISSED** [review]

File: be/src/exec/pipeline/scan/morsel.cpp (hunk 11 above)
Comment (ZiheLiu) [reply be/src/exec/pipeline/scan/morsel.cpp:479]: Yes, if there isn't rowset, `_segment_group` is nullptr, and will go forward next rowset in `try_get`.

File: be/src/exec/pipeline/scan/morsel.cpp (hunk 12 above)
Comment (ZiheLiu) [reply be/src/exec/pipeline/scan/morsel.cpp:288]: Great.

**Review by murphyatwork — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**